### PR TITLE
fozzie-components@v3.36.0 - Adding Url builder to tests #globalconfig

### PR DESCRIPTION
**Pull request by Festivejelly**
This PR moves the set up of knobs from within the page object models and introduces a URL builder.

With the URL builder you can specify knobs as query strings of a URL.

This keeps Storybook specific code and logic out of the page object models.
It also makes it much easier to see which mocks a test is using from looking at the specfile directly.

Adding knobs can be setup like so:

`checkout.withQuery('&knob-Service Type', 'collection')
                .withQuery('&knob-Is User Logged In', true);
const pageUrl = buildUrl(checkout.componentType, checkout.componentName, checkout.path);
checkout.open(pageUrl)`

An improvement to be made in future is to not pass every param into buildUrl but pass the component itself. This would require pushing a new package version.

### CODE REVIEW

File: packages/components/organisms/f-cookie-banner/test-utils/component-objects/f-cookieBanner.component.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const Page = require('@justeat/f-wdio-utils/src/page.object');
```
Comment (siggerzz) [inline packages/components/organisms/f-cookie-banner/test-utils/component-objects/f-cookieBanner.component.js:1]: I don't think this file is actually used anywhere, so can be deleted.

File: packages/components/organisms/f-cookie-banner/test-utils/component-objects/f-cookieConsentBanner.component.js [hunk 2]
```diff
@@ -1,74 +1,55 @@
...
             case 'full':
                 this.cookieAcceptAllButton.click();
-            break;
+                break;
             case 'necessary':
                 this.cookieAcceptNecessaryButton.click();
-            break;
+                break;
+            default:
+                    // not implemented
```
Comment (siggerzz) [inline packages/components/organisms/f-cookie-banner/test-utils/component-objects/f-cookieConsentBanner.component.js]: Throw an exception here.

File: packages/components/organisms/f-cookie-banner/test/specs/component/desktop/f-cookieBanner-legacy.component.spec.js [hunk 3]
```diff
@@ -28,7 +32,12 @@ describe('Legacy - Multi-tenant - f-cookieBanner component tests', () => {
     .it('should go to the correct cookie policy page', (tenant, expectedCookiePolicyUrl) => {
 
         // Arrange
-        cookieBanner.open(tenant);
+        const countryFormatted = tenant.toUpperCase();
+        const formattedLocale = `en-${countryFormatted}`;
```
Comment (siggerzz) [inline packages/components/organisms/f-cookie-banner/test/specs/component/desktop/f-cookieBanner-legacy.component.spec.js]: Could simplify this to `const formattedLocale = en-${tenant.toUpperCase()};`

File: packages/components/organisms/f-cookie-banner/test/specs/component/desktop/f-cookieConsentBanner.component.spec.js [hunk 4]
```diff
@@ -34,9 +39,33 @@ describe('New - Multi-tenant - f-cookieBanner component tests', () => {
         // ['no', 'no/informasjonskapselerklaering']  'dk' and 'no' disabled for now
     ])
     .it('should go to the correct cookie policy page', (tenant, expectedCookiePolicyUrl) => {
-
         // Arrange
-        cookieBanner.open(tenant);
+        const countryFormatted = tenant.toUpperCase();
```
Comment (siggerzz) [inline packages/components/organisms/f-cookie-banner/test/specs/component/desktop/f-cookieConsentBanner.component.spec.js]: It'd make this a forEach() test and pass the full tenant code (e.g en-IE / da-DK /  it-IT / nb-NO).

Will tidy the test up a bit 👍🏼

File: packages/components/organisms/f-cookie-banner/test/specs/component/shared/f-cookieBanner-legacy.component.spec.js [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
+const cookieBanner = new CookieBanner('organism', 'cookie-banner-component');
+
 describe('Legacy - Multi-tenant - f-cookieBanner component tests', () => {
     forEach(['gb', 'au', 'nz'])
     .it('should display the f-cookieBanner component', tenant => {
         // Arrange
-        cookieBanner.open(tenant);
-        cookieBanner.waitForComponent
+        const countryFormatted = tenant.toUpperCase();
+        const formattedLocale = `en-${countryFormatted}`;
```
Comment (siggerzz) [inline packages/components/organisms/f-cookie-banner/test/specs/component/shared/f-cookieBanner-legacy.component.spec.js]: Simplify to `const formattedLocale = en-${tenant.toUpperCase()};`

File: packages/components/organisms/f-cookie-banner/test/specs/component/shared/f-cookieConsentBanner.component.spec.js [hunk 6]
```diff
@@ -26,9 +31,34 @@ describe('New - f-cookieBanner component tests', () => {
 
 describe('New - Multi-tenant - f-cookieBanner component tests', () => {
     forEach(['es', 'ie', 'it']) // 'dk' and 'no' disabled for now
-    .it.only('should display the f-cookieBanner component', (tenant) => {
+    .it.only('should display the f-cookieBanner component', tenant => {
```
Comment (siggerzz) [inline packages/components/organisms/f-cookie-banner/test/specs/component/shared/f-cookieConsentBanner.component.spec.js]: Remove `.only`

You should make the `tenant` param the full tenant code, then we can remove the switch 👍🏼

File: packages/components/organisms/f-footer/test/specs/component/shared/f-footer.component.spec.js [hunk 7]
```diff
@@ -1,17 +1,46 @@
 const forEach = require('mocha-each');
+
+const { buildUrl } = require('@justeat/f-wdio-utils/src/storybook-extensions.js');
 const Footer = require('../../../../test-utils/component-objects/f-footer.component');
 
-const footer = new Footer();
+const footer = new Footer('organism', 'footer-component');
+
+function formatLocale (tenant) {
```
Comment (siggerzz) [inline packages/components/organisms/f-footer/test/specs/component/shared/f-footer.component.spec.js]: Ditto.

Remove this function and just change the tests to the full tenant code

File: packages/components/organisms/f-header/test/specs/component/desktop/f-header.component.spec.js [hunk 8]
```diff
@@ -1,16 +1,46 @@
 const forEach = require('mocha-each');
+
+const { buildUrl } = require('@justeat/f-wdio-utils/src/storybook-extensions.js');
 const Header = require('../../../../test-utils/component-objects/f-header.component');
-const header = new Header();
+
+const header = new Header('organism', 'header-component');
+
+function formatLocale (tenant) {
```
Comment (siggerzz) [inline packages/components/organisms/f-header/test/specs/component/desktop/f-header.component.spec.js]: ditto

File: packages/components/organisms/f-header/test/specs/component/mobile/f-header.component.spec.js [hunk 9]
```diff
@@ -1,16 +1,46 @@
 const forEach = require('mocha-each');
+
+const { buildUrl } = require('@justeat/f-wdio-utils/src/storybook-extensions.js');
 const Header = require('../../../../test-utils/component-objects/f-header.component');
-const header = new Header();
+
+const header = new Header('organism', 'header-component');
+
+function formatLocale (tenant) {
```
Comment (siggerzz) [inline packages/components/organisms/f-header/test/specs/component/mobile/f-header.component.spec.js]: ditto

**Review by siggerzz — CHANGES_REQUESTED** [review]
Just a few minor changes!

**Comment by Festivejelly** [comment]
Valid feedback on the PR but I didnt create most of these tests. I dont want to get into a loop of doing lots of improvements. My PR is just to add the URL functionality. Ill change it based on the current feedback but please bear in mind that these tests are not authored by me so ideally should be refactored separately.

File: packages/components/atoms/f-button/test/specs/component/f-button.component.spec.js [hunk 10]
```diff
@@ -1,21 +1,32 @@
+const { buildUrl } = require('@justeat/f-wdio-utils/src/storybook-extensions.js');
 const Button = require('../../../test-utils/component-objects/f-button.component');
-const button = new Button();
+
+const button = new Button('atom-folder', 'f-button--button-component');
 
 describe('f-button component tests', () => {
 
     it('should display the f-button action component', () => {
+        // Arrange
```
Comment (LTurns) [inline packages/components/atoms/f-button/test/specs/component/f-button.component.spec.js]: Are we changing line 10 to only pass in the component, like `buildUrl(button)`? I think you mentioned this before so may already be doing it 😄

Comment (Festivejelly) [reply packages/components/atoms/f-button/test/specs/component/f-button.component.spec.js]: Correct but that requires changing the webdriver io utils function, which ill do as a different PR, then ill create a different PR to refactor the tests to just take in the 1 param.

Comment (LTurns) [reply packages/components/atoms/f-button/test/specs/component/f-button.component.spec.js]: awesome, sounds good!

**Comment by ashleynolan** [comment]
Small one – but should use the `#globalconfig` flag rather than the trivial flag. That way we still add a `CHANEGLOG` entry to the base repo for a change like this (which is quite big, so it's good to have that tracked somewhere).

File: packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js [hunk 11]
```diff
@@ -26,10 +31,15 @@ describe('New - f-cookieBanner component tests - @browserstack', () => {
     });
 
     // 'dk' and 'no' disabled for now
-    forEach(['es', 'ie', 'it'])
+    forEach(['es-ES', 'en-IE', 'it-IT'])
     .it('should display the f-cookieBanner component for "%s"', tenant => {
         // Arrange
-        cookieBanner.open(tenant);
+        cookieBanner = new CookieBanner('organism', 'cookie-banner-component');
```
Comment (JoshuaNg2332) [inline packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js:37]: Do we need this `cookieBanner` when we're already calling it in the top level describe?

File: packages/components/organisms/f-header/test/accessibility/axe-accessibility.spec.js [hunk 12]
```diff
@@ -1,25 +1,24 @@
...
-        const axeResults = getAccessibilityTestResults('f-header');
-
-        // Assert
-        expect(axeResults.violations.length).toBe(0);
-    });
-});
+//         // Assert
+//         expect(axeResults.violations.length).toBe(0);
+//     });
+// });
```
Comment (JoshuaNg2332) [inline packages/components/organisms/f-header/test/accessibility/axe-accessibility.spec.js]: Commented code?

File: packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js (hunk 11 above)
Comment (LTurns) [reply packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js:37]: ahh yeah we do - I've just checked and it errors if I delete the new instance (just line 37) sadly

File: packages/components/organisms/f-header/test/accessibility/axe-accessibility.spec.js (hunk 12 above)
Comment (LTurns) [reply packages/components/organisms/f-header/test/accessibility/axe-accessibility.spec.js]: whoops thank you! Will fix now

File: packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js (hunk 11 above)
Comment (JoshuaNg2332) [reply packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js:37]: Ok cool. I've seen a couple instances of this in other tests and wanted to check if this was right. All good then. :)

Comment (Festivejelly) [reply packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js:37]: It needs a new instance to override the old one that's in memory from another test.

Comment (Festivejelly) [reply packages/components/organisms/f-cookie-banner/test/component/f-cookieConsentBanner.component.shared.spec.js:37]: else you get conflicts with URL arguments that have different values (knobs)

**Review by JoshuaNg2332 — DISMISSED** [review]

**Review by siggerzz — APPROVED** [review]
There's a bit of minor refactoring that can be done to this, but given how long this ticket has been open, I'm happy to address this in another PR.

**Review by JoshuaNg2332 — APPROVED** [review]

**Pull request merged**